In [2]:
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd

In [3]:
df=pd.read_csv('flights.csv')

In [4]:
df.head()

YEAR  MONTH  DAY  DAY_OF_WEEK AIRLINE  FLIGHT_NUMBER TAIL_NUMBER  \
0  2015      5   16            6      UA           1511      N18220   
1  2015      4    7            2      AA            343      N3ALAA   
2  2015      1    8            4      OO           2640      N988CA   
3  2015      5    5            2      OO           5219      N971SW   
4  2015      9   22            2      DL            134      N3736C   

  ORIGIN_AIRPORT DESTINATION_AIRPORT  SCHEDULED_DEPARTURE  ...  ARRIVAL_TIME  \
0            PHX                 ORD                 1413  ...        1929.0   
1            LGA                 ORD                 1630  ...        1908.0   
2            ORD                 ICT                  835  ...        1312.0   
3            SBN                 ORD                  622  ...         555.0   
4            PHX                 DTW                 1003  ...        1638.0   

   ARRIVAL_DELAY  DIVERTED  CANCELLED  CANCELLATION_REASON  AIR_SYSTEM_DELAY  \
0          -11.0         0          0                  NaN               NaN   
1           48.0         0          0                  NaN              14.0   
2          157.0         0          0                  NaN               0.0   
3          -11.0         0          0                  NaN               NaN   
4          -11.0         0          0                  NaN               NaN   

   SECURITY_DELAY  AIRLINE_DELAY  LATE_AIRCRAFT_DELAY  WEATHER_DELAY  
0             NaN            NaN                  NaN            NaN  
1             0.0            0.0                 34.0            0.0  
2             0.0            0.0                  6.0          151.0  
3             NaN            NaN                  NaN            NaN  
4             NaN            NaN                  NaN            NaN  

[5 rows x 31 columns]

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5819 entries, 0 to 5818
Data columns (total 31 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   YEAR                 5819 non-null   int64  
 1   MONTH                5819 non-null   int64  
 2   DAY                  5819 non-null   int64  
 3   DAY_OF_WEEK          5819 non-null   int64  
 4   AIRLINE              5819 non-null   object 
 5   FLIGHT_NUMBER        5819 non-null   int64  
 6   TAIL_NUMBER          5808 non-null   object 
 7   ORIGIN_AIRPORT       5819 non-null   object 
 8   DESTINATION_AIRPORT  5819 non-null   object 
 9   SCHEDULED_DEPARTURE  5819 non-null   int64  
 10  DEPARTURE_TIME       5740 non-null   float64
 11  DEPARTURE_DELAY      5740 non-null   float64
 12  TAXI_OUT             5739 non-null   float64
 13  WHEELS_OFF           5739 non-null   float64
 14  SCHEDULED_TIME       5819 non-null   int64  
 15  ELAPSED_TIME         5719 non-null   f

1/ this dataset has 5819 rows and 30 columns, it contains information about flights in and   out of the USA in a period of time we'll get to know later .

In [6]:
df.isna().sum()

YEAR                      0
MONTH                     0
DAY                       0
DAY_OF_WEEK               0
AIRLINE                   0
FLIGHT_NUMBER             0
TAIL_NUMBER              11
ORIGIN_AIRPORT            0
DESTINATION_AIRPORT       0
SCHEDULED_DEPARTURE       0
DEPARTURE_TIME           79
DEPARTURE_DELAY          79
TAXI_OUT                 80
WHEELS_OFF               80
SCHEDULED_TIME            0
ELAPSED_TIME            100
AIR_TIME                100
DISTANCE                  0
WHEELS_ON                84
TAXI_IN                  84
SCHEDULED_ARRIVAL         0
ARRIVAL_TIME             84
ARRIVAL_DELAY           100
DIVERTED                  0
CANCELLED                 0
CANCELLATION_REASON    5738
AIR_SYSTEM_DELAY       4727
SECURITY_DELAY         4727
AIRLINE_DELAY          4727
LATE_AIRCRAFT_DELAY    4727
WEATHER_DELAY          4727
dtype: int64

we have 11 airplane TAIL_NUMBER missing, and 79 flight departure_time and departure_delay ,this means that the plane never departed in other words the flight was canceled , we also have approxamatly the same number of NAs in (TAXI_OUT,WHEELS_OFF,ELAPSED_TIME,AIR_TIME,WHEELS_ON,TAXI_IN,ARRIVAL_TIME,ARRIVAL_DELAY) probably for the same reason but we'll know more when we analyze each variable on it's own , we have 4727 missing values in the delay reasons variables , this means that there was no delay on these flights , and 5738 missing values in the canncellation reason , this means that around 5738 flight weren't canceled 

# Uni variate analysis :

## Numeric variables 